usagov/usagov-2021
# Usagov 1296 recaptcha error style

## Jira Task

https://cm-jira.usa.gov/browse/USAGOV-1296

## Description

The main goal of this change was to make the reCaptcha box responsive and fix the reCaptcha error style. The following changes were made to the code:
- Add a new reCaptcha so that we have two reCaptchas, 1 for smaller devices (screen size <= 499px) and 1 for larger devices (screen size >= 500px). _Note: I tried other options, such as changing the "size" attribute of the reCaptcha depending on the screen size, but it didn't work._
- Add a listener function on state changes so that runs the new function `changeRecaptchaDisplay` every time the screen changes from 499px to 500px and viceversa. 
- Add a MutationObserver that runs the functions `modifyErrorElements` and `reCaptchaValidation` every time the attribute "style" changes in the large reCaptcha. Basically, this validates if the reCaptcha response, adds the error style if needed, and modifies the alert box.

It's important to mention that if the user validates the small reCaptcha and then changes the size of the reCaptcha, they will have to re-validate it since it's not possible to share the validation between two reCaptchas.

## Type of Changes

  - Add screenshot showing what it should look like

## Testing Instructions


1. Clear all caches.
2. Make sure that you're screen size is >= 500px.
3. Go to localhost/site-issue-report-form and localhost/es/reporte-problemas-en-este-sitio-web and check that it looks like this (the reCaptcha should be in the center when it does not have the error message or the outline):

English Version:
<img width="437" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/6cf5dfeb-c1d5-49f0-b2b8-61453b1abf90">

Spanish Version:
<img width="439" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/a6967418-0a78-4078-bb23-ee97c10d7dbf">


4. In both languages, leave all fields blank and submit the form. Inspect the page and check that the href of reCaptcha link inside the alert box is "#recaptcha-large-container". You should see the following error messages and style (the reCaptcha must be left aligned when it has the border and error message):

English Version:
<img width="1192" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/f0665969-4e70-4443-9bf2-b7ba30f7ef23">

Spanish Version:
<img width="1187" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/1f7dcc80-f71f-4646-8c3a-25d36e96a9fe">

5. Change the screen size to <= 499px.
6. In both languages, check that it looks like this (the reCaptcha should be in the center when it does not have the error message or the outline and the data-size="compact"):

English Version:
<img width="1303" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/6bb66e27-ab70-4766-90d7-9baec7a56452">

Spanish Version:
<img width="1303" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/e36b129b-f15a-4b14-aaa8-75661b104a70">


7. In both languages, leave all fields blank and submit the form. Inspect the page and check that the href of reCaptcha link inside the alert box is "#recaptcha-small-container". You should see the following error messages and style (the reCaptcha must be left aligned when it has the border and error message):

English Version:
<img width="1301" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/60c552f6-c0d4-47e2-bfa0-0681f0a7f2b3">
<img width="407" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/dbdaeb8a-1498-4eef-94c9-f5bae647648d">

Spanish Version:
<img width="1292" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/624770e8-8733-4653-b6bb-4263ec3cf15f">
<img width="385" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/f18c6487-88f8-49e6-a704-1676c45c5efa">


### Requires New Config

### Requires New Content

### Validation Steps

## Security Review




## Reviewer Reminders
- Reviewed code changes
- Reviewed functionality
- Security review complete or not required

## Post PR Approval Instructions
Follow these steps as soon as you merge the new changes.

1. Go to the [USAGov Circle CI project](https://app.circleci.com/pipelines/github/usagov/usagov-2021).
2. Find the commit of this pull request.
3. Build and deploy the changes.
8. Update the Jira ticket by changing the ticket status to `Review in Test` and add a comment. State whether the change is already visible on [cms-dev.usa.gov](http://cms-dev.usa.gov/) and [beta-dev.usa.gov](http://beta-dev.usa.gov/), or if the deployment is still in process.
